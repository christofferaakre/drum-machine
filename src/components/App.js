import React from 'react';

import Slider from 'react-rangeslider';

import Header from './Header';
import Audio from './Audio';
import audio from '../utils/audio';
import theme from '../theme';

export const styles = {
  container: {
    display       : 'flex',
    justifyContent: 'center',
    alignItems    : 'center',
    height        : '100vh',
    padding       : 25,
  },
  gridContainer: {
    display            : 'grid',
    gridTemplateColumns: '1fr 1fr 1fr',
    gridGap            : 15,
  },
  display: {
    display      : 'flex',
    flexDirection: 'column',
    alignItems   : 'center',
    padding      : 15,
    width        : 200,
  },
  lastPlayed: {
    color   : theme.palette.primary.main,
    fontSize: 28,
    padding : 15,
  },
  sliderContainer: {
    width  : 200,
    padding: '0 25px',
  },
};

export class App extends React.Component {
  state = {
    volume    : 0.5,
    lastPlayed: '',
  };
  componentDidMount() {
    let sounds = {};

    audio.forEach(({ keyToPress, }) => {
      const Key = keyToPress.toUpperCase();

      sounds = { ...sounds, [Key]: document.getElementById(Key), };
    });
    document.addEventListener('keypress', ({ key, }) => {
      const Key = key.toUpperCase();

      if (audio.map(({ keyToPress, }) => keyToPress).includes(Key)) {
        sounds[Key].play();
      }
    });
  }
  handleVolumeChange(volume) {
    this.setState(() => ({ volume: Math.round(volume) / 100, }));
  }
  handlePlay = name => this.setState(() => ({ lastPlayed: name, }));
  render() {
    const { volume, lastPlayed, } = this.state;

    return (
      <div id='container'>
        <Header />
        <div style={styles.container} id='drum-machine'>
          <div style={styles.gridContainer} id='drum-pads-container'>
            {audio.map(({ src, keyToPress, name, }) => (
              <Audio
                key={src}
                src={src}
                name={name}
                keyToPress={keyToPress}
                volume={volume}
                handlePlay={this.handlePlay}
              />
            ))}
          </div>
          <div style={styles.display} id='display'>
            <span style={styles.lastPlayed}>{lastPlayed || 'Click a drum'}</span>
            <div style={styles.sliderContainer} id='slider-container'>
              <Slider
                id='volume-slider'
                min={0}
                max={100}
                value={Math.round(volume * 100)}
                onChange={value => this.handleVolumeChange(value)}
              />
            </div>
            <span style={styles.volume}>Volume: {Math.round(volume * 100)}%</span>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
