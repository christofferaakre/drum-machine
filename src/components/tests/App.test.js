import React from 'react';
import { shallow, } from 'enzyme';
import { App, } from '../App';

let wrapper, props;

beforeEach(() => {
  wrapper = shallow(<App {...props} />);
});

describe('App', () => {
  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe('componentDidMount', () => {
    test('adds event listener to keypress', () => {
      document.addEventListener = jest.fn();
      wrapper.instance().componentDidMount();
      expect(document.addEventListener).toHaveBeenCalled();
    });
    test();
  });
  test('calls handleVolumeChange when slider value changes', () => {
    const spy = jest.spyOn(App.prototype, 'handleVolumeChange');

    wrapper.find('#volume-slider').simulate('change');
    expect(spy).toHaveBeenCalled();
  });
  describe('handleVolumeChange', () => {
    test('calls setState', () => {
      const setState = jest.spyOn(App.prototype, 'setState');

      wrapper.instance().handleVolumeChange(50);
      expect(setState).toHaveBeenCalled();
    });
    test('sets the volume state', () => {
      wrapper.instance().handleVolumeChange(50);
      expect(wrapper.state('volume')).toBe(0.5);
    });
  });
  describe('handlePlay', () => {
    test('calls setState', () => {
      const setState = jest.spyOn(App.prototype, 'setState');

      wrapper.instance().handlePlay('Clap');
      expect(setState).toHaveBeenCalled();
    });
    test('sets the lastPlayed state', () => {
      wrapper.instance().handlePlay('Clap');
      expect(wrapper.state('lastPlayed')).toBe('Clap');
    });
  });
});
