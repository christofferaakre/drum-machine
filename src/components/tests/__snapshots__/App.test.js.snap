// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  id="container"
>
  <Header />
  <div
    id="drum-machine"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100vh",
        "justifyContent": "center",
        "padding": 25,
      }
    }
  >
    <div
      id="drum-pads-container"
      style={
        Object {
          "display": "grid",
          "gridGap": 15,
          "gridTemplateColumns": "1fr 1fr 1fr",
        }
      }
    >
      <Audio
        handlePlay={[Function]}
        key="audio/clap.wav"
        keyToPress="Q"
        name="Clap"
        src="audio/clap.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/crash.wav"
        keyToPress="W"
        name="Crash"
        src="audio/crash.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/hihat.wav"
        keyToPress="E"
        name="Hi-hat"
        src="audio/hihat.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/kick.wav"
        keyToPress="A"
        name="Kick"
        src="audio/kick.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/openhat.wav"
        keyToPress="S"
        name="Open-hat"
        src="audio/openhat.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/snare-1.wav"
        keyToPress="D"
        name="Snare 1"
        src="audio/snare-1.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/snare-2.wav"
        keyToPress="Z"
        name="Snare 2"
        src="audio/snare-2.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/tom-1.wav"
        keyToPress="X"
        name="Tom 1"
        src="audio/tom-1.wav"
        volume={0.5}
      />
      <Audio
        handlePlay={[Function]}
        key="audio/tom-2.wav"
        keyToPress="C"
        name="Tom 2"
        src="audio/tom-2.wav"
        volume={0.5}
      />
    </div>
    <div
      id="display"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "padding": 15,
          "width": 200,
        }
      }
    >
      <span
        style={
          Object {
            "color": "#009688",
            "fontSize": 28,
            "padding": 15,
          }
        }
      >
        Click a drum
      </span>
      <div
        id="slider-container"
        style={
          Object {
            "padding": "0 25px",
            "width": 200,
          }
        }
      >
        <Slider
          handleLabel=""
          id="volume-slider"
          labels={Object {}}
          max={100}
          min={0}
          onChange={[Function]}
          orientation="horizontal"
          reverse={false}
          step={1}
          tooltip={true}
          value={50}
        />
      </div>
      <span>
        Volume: 
        50
        %
      </span>
    </div>
  </div>
</div>
`;
